Archipelago.HollowKnight/Goals.cs

```diff
@@ -0,0 +1,172 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+using UnityEngine.SceneManagement;
+using ItemChanger;
+
+namespace Archipelago.HollowKnight
+{
+    public enum Goals
+    {
+        Any = 0,
+        HollowKnight = 1,
+        SealedSiblings = 2,
+        Radiance = 3,
+        Godhome = 4,
+        MAX = Godhome
+    }
+
+    abstract class Goal {
+        private static readonly Dictionary<Goals, Goal> Lookup = new Dictionary<Goals, Goal>()
+        {
+            [Goals.Any] = new AnyGoal(),
+            [Goals.HollowKnight] = new HollowKnightGoal(),
+            [Goals.SealedSiblings] = new SealedSiblingsGoal(),
+            [Goals.Radiance] = new RadianceGoal(),
+            [Goals.Godhome] = new GodhomeGoal()
+        };
+
+        public static Goal GetGoal(Goals key)
+        {
+            Goal value;
+            if (Lookup.TryGetValue(key, out value))
+            {
+                return value;
+            }
+            return null;
```

Comment:
Throw an exception here, something isn't right if we get a `Goals` value that doesn't correspond to a `Goal` in the `Lookup` object.